lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+# frozen_string_literal: true
+
+module RGeo
+  module Cartesian
+    module Graphs
+      # HalfEdge represents an edge as 2 directed edges.
+      # One half-edge will have it's origin at edge.s, the other
+      # at edge.e. Both half-edges will be linked as each other's twins.
+      #
+      # HalfEdges also contain references to the next and prev half-edges,
+      # where next's origin is this half-edges destination. Prev's destination
+      # is this half-edge's origin.
+      class HalfEdge
+        include Comparable
+
+        # Creates 2 half edges from an edge.
+        # They will be assigned as each other's twins.
+        # The Half Edges will be returned in the order of points
+        # of the edge (start, end).
+        #
+        # @param edge [RGeo::Cartesian::Segment]
+        #
+        # @return [Array]
+        def self.from_edge(edge)
+          # TODO: should we raise an error if the edge is degenerate?
+          e1 = new(edge.s)
+          e2 = new(edge.e)
+
+          e1.twin = e2
+          e2.twin = e1
+          [e1, e2]
+        end
+
+        def initialize(origin)
+          @origin = origin
+          @twin = nil
+          @next = nil
+          @prev = nil
+        end
+        attr_reader :origin
+        attr_accessor :twin, :next, :prev
+
+        # HalfEdges will be sorted around their shared vertex
+        # in a CW fashion. This means that face interiors will be
+        # a CCW.
+        def <=>(other)
+          angle <=> other.angle
+        end
+
+        # Will keep following next in a half-edge until it returns to itself
+        # or hits a half-edge without a next edge.
+        #
+        # @return [Array]
+        def each
+          hedges = []
+          yield(self) if block_given?
+          hedges << self
+
+          n = self.next
+          until n.eql?(self) || n.nil?
+            yield(n) if block_given?
+            hedges << n
+            n = n.next
+          end
+          hedges
+        end
+
+        # Return the destination of the half edge
+        #
+        # @return [RGeo::Feature::Point]
+        def destination
+          twin.origin
```

Comment:
```suggestion
          twin&.origin
```

This will help you later, for instance in the `#to_s` method